meteor/server/publications/mountedTriggers.ts

```diff
@@ -0,0 +1,121 @@
+import { Meteor } from 'meteor/meteor'
+import { CustomPublish, CustomPublishChanges, meteorCustomPublish } from '../lib/customPublication'
+import { CustomCollectionName, PubSub } from '../../lib/api/pubsub'
+import { PeripheralDeviceId } from '@sofie-automation/corelib/dist/dataModel/Ids'
+import { PeripheralDeviceReadAccess } from '../security/peripheralDevice'
+import { PeripheralDevices } from '../../lib/collections/PeripheralDevices'
+import { logger } from '../logging'
+import { DeviceTriggerMountedActionAdlibsPreview, DeviceTriggerMountedActions } from '../api/deviceTriggers/observer'
+import { Mongo } from 'meteor/mongo'
+import { ProtectedString } from '@sofie-automation/corelib/dist/protectedString'
+import _ from 'underscore'
+
+const PUBLICATION_DEBOUNCE = 20
+
+meteorCustomPublish(
+	PubSub.mountedTriggersForDevice,
+	CustomCollectionName.MountedTriggers,
+	async function (pub, deviceId: PeripheralDeviceId, deviceIds: string[], token) {
+		if (await PeripheralDeviceReadAccess.peripheralDeviceContent(deviceId, { userId: this.userId, token })) {
+			const peripheralDevice = PeripheralDevices.findOne(deviceId)
+
+			if (!peripheralDevice) throw new Meteor.Error(404, `PeripheralDevice "${deviceId}" not found`)
+
+			const studioId = peripheralDevice.studioId
+			if (!studioId) throw new Meteor.Error(400, `Peripheral Device "${deviceId}" not attached to a studio`)
+
+			cursorCustomPublish(
+				pub,
+				DeviceTriggerMountedActions.find({
+					studioId,
+					deviceId: {
+						$in: deviceIds,
+					},
+				})
+			)
+		} else {
+			logger.warn(`Pub.mountedTriggersForDevice: Not allowed: "${deviceId}"`)
+		}
+	}
+)
+
+meteorCustomPublish(
+	PubSub.mountedTriggersForDevicePreview,
+	CustomCollectionName.MountedTriggersPreviews,
+	async function (pub, deviceId: PeripheralDeviceId, token) {
+		if (await PeripheralDeviceReadAccess.peripheralDeviceContent(deviceId, { userId: this.userId, token })) {
+			const peripheralDevice = PeripheralDevices.findOne(deviceId)
+
+			if (!peripheralDevice) throw new Meteor.Error(404, `PeripheralDevice "${deviceId}" not found`)
+
+			const studioId = peripheralDevice.studioId
+			if (!studioId) throw new Meteor.Error(400, `Peripheral Device "${deviceId}" not attached to a studio`)
+
+			cursorCustomPublish(
+				pub,
+				DeviceTriggerMountedActionAdlibsPreview.find({
+					studioId,
+				})
+			)
+		} else {
+			logger.warn(`Pub.mountedTriggersForDevicePreview: Not allowed: "${deviceId}"`)
+		}
+	}
+)
+
+function cursorCustomPublish<T extends { _id: ProtectedString<any> }>(pub: CustomPublish<T>, cursor: Mongo.Cursor<T>) {
+	let buffer: CustomPublishChanges<T> = {
```

Comment:
How clustered are these updates usually? I am wondering if it would be better to have the buffer be formed as
```
let buffer = {
  added: new Map(),
  changed: new Map(),
  removed: new Set()
}
```
and converted back to `CustomPublishChanges` inside the debounce.

That will avoid needing so much array searching in the observe callbacks. The cost of converting back might be too high to make it worthwhile

Reply:
They will be generally very clustered. I think it is probably a good idea to convert these to Sets & Maps.